src/dynamic_instruction.rs

```diff
@@ -0,0 +1,14 @@
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub enum DynamicInstruction {
```

Comment:
It would be nice to generalize the file and buffer loading, perhaps something like:
1. `StoreFromFile(String)` and `Store(offset, data, final)` instructions that are used to fill up an `Account` with the program binary be it native, bpf, lua, wasm, etc.  (That `final` arg to `Store` marks the Account as finished, so no future `Store` instructions can write into it). 

2. Then we have `LoadBfp`/`LoadNative`/`LoadXXX` instructions that accept the `Account` pubkey that was just filled with `Store`(s).   This set of instructions validate/load the code in the `Account`, and if successful set the right bits/data structures in the bank so that future `Transactions` can target the `Account` pubkey as their `program_id` (which means we need no `Call` instruction).

Reply:
One thing to keep in mind is that the account that was loaded only contains the raw bits.  There may be other things that need to be done on a per call basis (increment an underlying reference count, a new instance of the vm, etc...  Aka, the Loads you are talking about in point 2 above are actually what needs to happen in the call and hence a separate instruction.

If we follow what you suggest I think we'll need an additional account.  An account to hold the raw bits and another to hold the 'loaded' program (handle, vm instance (assuming its reentrant), etc...).

Reply:
Are you expecting Solana to re-alloc that program account for the user?

Reply:
Let's define two things first:
1. The user-visible `Account` with the raw bits.  This `Account` is assigned to the `dynamic_program`'s `program_id`, but is still read-able by all (say if you wanted to inspect the program bits before using it).
2. Internal Solana data structures, VM specific, that turn those raw bits into something that can respond to incoming `Transaction` `Instruction`s.  These are callable by bank.rs, like what we currently do at https://github.com/solana-labs/solana/blob/master/src/bank.rs#L443-L464.   There's no `DynamicInstruction::Call` needed here, the `Transaction` `Instruction` userdata is simply routed to the entrypoint of the program and it figures out what to do.

Depending on the interpreter, getting from (1) to (2) may require a re-alloc internally (not visible to the user of the system).

Reply:
I shoved that bank.rs function in there but don't like it.  That was based on a native module assumption.  Take a look at dynamic_program.rs's implementation of Call.  There you can see what kinds if things need to be done on call.  For example, a new instance of the VM for bpf programs.  Curious if how you see those things moved from the proposed Call instruction.

Reply:
Heh, that's funny.  The current bank.rs abstraction seems ideal to me from the user perspective.  Two things I see in dynamic_program's Call implementation:

1.  We're incurring a 4 byte charge in every Call's userdata.  Rust enum serializes down to 4 bytes, we could make that 1 byte with a custom serde but it's not zero -- and it could be zero if we continue to shove stuff in bank.rs (with perhaps a more appealing interface/abstraction).   

2. Current Call loads the program on every transaction (except for the TLS workaround on Linux), this overhead would be prohibitive in a real system and we'll need a LRU cache of frequently invoked programs.  Once we build that LRU to improve performance we'll probably have something that looks like the  "2. Internal Solana data structures" I mention above.